vnext/Shared/JSI/NapiJsiRuntime.cpp

```diff
@@ -0,0 +1,1921 @@
+}
+
+// Creates a new JavaScript Array with the provided length.
+napi_value NapiJsiRuntime::CreateArray(size_t length) const {
+  napi_value result{};
+  CHECK_NAPI(napi_create_array_with_length(m_env, length, &result));
+
+  return result;
+}
+
+// Sets array element.
+void NapiJsiRuntime::SetElement(napi_value array, uint32_t index, napi_value value) const {
+  CHECK_NAPI(napi_set_element(m_env, array, index, value));
+}
+
+// The NAPI external function callback used for the JSI host function implementation.
+/*static*/ napi_value NapiJsiRuntime::JsiHostFunctionCallback(napi_env env, napi_callback_info info) noexcept {
+  HostFunctionWrapper *hostFuncWrapper{};
+  size_t argc{};
+  CHECK_NAPI_ELSE_CRASH(
+      napi_get_cb_info(env, info, &argc, nullptr, nullptr, reinterpret_cast<void **>(&hostFuncWrapper)));
+  CHECK_ELSE_CRASH(hostFuncWrapper, "Cannot find the host function");
+  NapiJsiRuntime &runtime = hostFuncWrapper->GetRuntime();
+
+  return runtime.HandleCallbackExceptions([&/*TODO: explicit captures*/]() {
+    SmallBuffer<napi_value, MaxStackArgCount> napiArgs(argc);
+    napi_value thisArg{};
+    CHECK_NAPI_ELSE_CRASH(napi_get_cb_info(env, info, &argc, napiArgs.Data(), &thisArg, nullptr));
+    CHECK_ELSE_CRASH(napiArgs.Size() == argc, "Wrong argument count");
+    const JsiValueView jsiThisArg{&runtime, thisArg};
+    JsiValueViewArgs jsiArgs(&runtime, span<napi_value>(napiArgs.Data(), napiArgs.Size()));
+
+    const HostFunctionType &hostFunc = hostFuncWrapper->GetHosFunction();
+    return runtime.RunInMethodContext("HostFunction", [&/*TODO: explicit captures*/]() {
+      return runtime.GetNapiValue(hostFunc(runtime, jsiThisArg, jsiArgs.Data(), jsiArgs.Size()));
+    });
+  });
+}
+
+// Creates an external function.
+napi_value NapiJsiRuntime::CreateExternalFunction(
+    napi_value name,
+    int32_t paramCount,
+    napi_callback callback,
+    void *callbackData) {
+  string funcName = StringToStdString(name);
+  napi_value function{};
+  CHECK_NAPI(napi_create_function(m_env, funcName.data(), funcName.length(), callback, callbackData, &function));
+  SetProperty(function, m_propertyId.length, CreateInt32(paramCount), napi_property_attributes::napi_default);
+
+  return function;
+}
+
+// Creates an object that wraps up external data.
+napi_value NapiJsiRuntime::CreateExternalObject(void *data, napi_finalize finalizeCallback) const {
+  napi_value result{};
+  CHECK_NAPI(napi_create_external(m_env, data, finalizeCallback, nullptr, &result));
+
+  return result;
+}
+
+// Wraps up std::unique_ptr as an external object.
+template <typename T>
+napi_value NapiJsiRuntime::CreateExternalObject(unique_ptr<T> &&data) const {
+  napi_value object =
+      CreateExternalObject(data.get(), [](napi_env /*env*/, void *dataToDestroy, void * /*finalizerHint*/) {
+        // We wrap dataToDestroy in a unique_ptr to avoid calling delete explicitly.
+        delete static_cast<T *>(dataToDestroy);
```

Comment:
need to ensure T is not an array type since you'd have to call `delete[]`  instead

Reply:
It is only used internally by the code in this .cpp file. We do not use T[] here.